content/blog/debugging-dde.md

```diff
@@ -0,0 +1,149 @@
+}
+if (a>0) printf("a>0\n");
+if (b>0) printf("b>0\n");
+if ((a>0) && (b>0)) printf("Both a>0 and b>0!");
+
+```
+...produces this output?
+```
+Before loop
+a>0
+b>0
+Both a>0 and b>0!
+```
+
+This, simplified, was a situation we were in when 
+[DDE](https://github.com/mrc-ide/dde) 0.9.0 was compiled, but only on Windows,
+and only when compiled with 32-bit gcc. With 64-bits, or linux, the loop
+behaved itself perfectly.
+
+But this is almost the end of the story. It began with a failing appveyor test
+on a trivial line of code. [DDE](https://github.com/mrc-ide/dde) is a package 
+that *solves ordinary differential equations (ODEs), 
+delay differential equations (DDEs) and discrete-time difference (or recursion) 
+equations, perhaps involving delays* - and all its tests and worked fine,
+except one. The test that produced different output for 32-bit than for 
+64-bit was about the simplest case possible: a function that always returns `1`.
+
+```
+  deriv <- function(t, y, p) { 1 }
+  dde::dopri(0, c(0, 1), deriv, 0)
+```
+
+For the two timesteps, `0` and `1`, on 64-bit, it correctly returned `(1,1)`; 
+on 32-bit, it returned `(1,0)`.
+
+## Primary Approaches
+
+At first, we wondered if it was a memory issue that only surfaced on 32-bit. 
+So we checked beyond doubt that `sizeof(size_t)` is `4` on 32-bit, but `8` on 
+64-bit, as are all pointers. All good. Other types were the same for all 
+platforms. But there were no obvious violations; Valgrind on linux picked up
+nothing.
+
+Resorting to random black-box bashing: `dde:dopri(0, c(0, 1.121), deriv, 0)`
+and smaller numbers failed the test. `dde:dopri(0, c(0, 1.122), deriov, 0)` and
+larger numbers all passed. Eureka? Nope - a mysterious red herring, alas.
+
+In the real-life code, the `(a>0)` and `(b>0)` were a bit more complicated -
+but not much; no functions involved, just comparisons between members of a
+struct, so no sensible way they could exit the loop, and then appear
+to have met the criteria for looping in retrospect, as we were seeing. 
+
+## Secondary Flailings
+
+The next level of desperation involved dumping every member of the object at
+intervals with `Rprintf`. This revealed some corruption at the end of the code, 
+surely evidence of some bad memory writes! Wrong again - they were tracked down
+to different, but very acceptable behaviour as the object had already been 
+freed and deconstructed by then, so could rightly return undefined values.
+
+But this did expose a genuine difference between platforms in a variable; 
+narrowed down to the `while` loop above, which terminated early (or didn't 
+get entered at all), under certain 32-bit circumstances. Printing things within
+the loop - or sometimes before it - sometimes changed the behaviour and made
+things work, but it would be poor form to submit a PR that claims to fix a functional
+problem by printing a ton of waffle before-hand.
+
+My github commit titles around this time are telling. **Total Confusion Reigns** 
```

Comment:
you can link to https://github.com/mrc-ide/dde/commits/i14_win32 if you want to

Reply:
Done - with slight embarrassment, but it does tell the story!